### PR TITLE
Make Sapling the default address format when calling RPC z_getnewaddress

### LINKED ISSUE #3671: Make "sapling" the default for RPC "z_getnewaddress"
Related to discussion in https://github.com/zcash/zcash/issues/3064 and recent offline discussion that it would be less confusing to users if `z_getnewaddress` returned a Sapling address rather than a Sprout address.

### PR DESCRIPTION
**Pull request by bitcartel**
Closes #3671.

### CODE REVIEW

**Comment by bitcartel** [comment]
Reviewers: Merge this PR last as other reviews also update tests and this PR will likely need to be rebased for those tests that have changed.

File: qa/rpc-tests/equihash_params.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, initialize_chain_clean, \
+    start_node, connect_nodes
+
+# Test equihash parameter change which occurs at network upgrade activation
+# To avoid time-out, this test uses N=96, K=5
+class EquihashUpgradeParamsTest(BitcoinTestFramework):
```
Comment (daira) [inline qa/rpc-tests/equihash_params.py]: This test doesn't seem to be related to the subject of the PR.

File: qa/rpc-tests/p2p-expiringsoontx.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, initialize_chain_clean, \
+    start_node, p2p_port
+from test_framework.blocktools import create_block, create_coinbase
+
+import os
+import time
+
+'''
+AcceptBlockTest -- test processing of unrequested blocks.
```
Comment (daira) [inline qa/rpc-tests/p2p-expiringsoontx.py]: This test doesn't seem to be related to the subject of the PR.

File: qa/rpc-tests/wallet_anchorfork.py.original [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# Copyright (c) 2016 The Zcash developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
+from time import *
+
+class WalletAnchorForkTest (BitcoinTestFramework):
```
Comment (daira) [inline qa/rpc-tests/wallet_anchorfork.py.original]: This test doesn't seem to be related to the subject of the PR.

File: qa/rpc-tests/wallet_large.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    assert_equal,
+    connect_nodes_bi,
+    initialize_chain_clean,
+    start_nodes,
+)
+
+from decimal import Decimal
+
+
+class LargeWalletTest(BitcoinTestFramework):
```
Comment (daira) [inline qa/rpc-tests/wallet_large.py]: This test doesn't seem to be related to the subject of the PR.

File: qa/rpc-tests/wallet_overwintertx.py.hack [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.authproxy import JSONRPCException
+from test_framework.util import assert_equal, initialize_chain_clean, \
+    start_nodes, connect_nodes_bi, sync_blocks, sync_mempools, \
+    wait_and_assert_operationid_status
+
+from decimal import Decimal
+
+class WalletOverwinterTxTest (BitcoinTestFramework):
```
Comment (daira) [inline qa/rpc-tests/wallet_overwintertx.py.hack]: This file appears to have been included by mistake, judging by the `.hack` suffix.

File: qa/rpc-tests/wallet_z_listreceivedbyaddress.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import connect_nodes_bi, initialize_chain_clean, start_nodes, stop_nodes, \
+    wait_and_assert_operationid_status, wait_bitcoinds
+
+from decimal import Decimal
+
+class WalletZListReceivedByAddressTest (BitcoinTestFramework):
```
Comment (daira) [inline qa/rpc-tests/wallet_z_listreceivedbyaddress.py]: This test doesn't seem to be related to the subject of the PR.

File: src/test/rpc_wallet_tests.cpp [hunk 7]
```diff
@@ -626,7 +626,7 @@ BOOST_AUTO_TEST_CASE(rpc_wallet_z_importexport)
     BOOST_CHECK(myaddrs == listaddrs);
 
     // Add one more address
-    BOOST_CHECK_NO_THROW(retValue = CallRPC("z_getnewaddress"));
+    BOOST_CHECK_NO_THROW(retValue = CallRPC("z_getnewaddress sprout"));
     std::string newaddress = retValue.get_str();
     auto address = DecodePaymentAddress(newaddress);
     BOOST_CHECK(IsValidPaymentAddress(address));
```
Comment (LarryRuane) [inline src/test/rpc_wallet_tests.cpp]: github doesn't let me comment on lines outside the narrow range of lines that include diffs, but I happened to notice these lines (starting at line 637) could be improved:
```
    // Check if too many args
    BOOST_CHECK_THROW(CallRPC("z_getnewaddress toomanyargs"), runtime_error);
```

File: qa/rpc-tests/equihash_params.py (hunk 1 above)
Comment (bitcartel) [reply qa/rpc-tests/equihash_params.py]: @daira Weird, was fine yesterday, let me check.

File: qa/rpc-tests/p2p-expiringsoontx.py (hunk 2 above)
Comment (bitcartel) [reply qa/rpc-tests/p2p-expiringsoontx.py]: force-pushed fix

File: qa/rpc-tests/wallet_large.py (hunk 4 above)
Comment (bitcartel) [reply qa/rpc-tests/wallet_large.py]: force-pushed fix

File: qa/rpc-tests/wallet_overwintertx.py.hack (hunk 5 above)
Comment (bitcartel) [reply qa/rpc-tests/wallet_overwintertx.py.hack]: force-pushed fix

File: qa/rpc-tests/wallet_z_listreceivedbyaddress.py (hunk 6 above)
Comment (bitcartel) [reply qa/rpc-tests/wallet_z_listreceivedbyaddress.py]: force-pushed fix

File: qa/rpc-tests/wallet_anchorfork.py.original (hunk 3 above)
Comment (bitcartel) [reply qa/rpc-tests/wallet_anchorfork.py.original]: force-pushed fix

File: qa/rpc-tests/wallet_addresses.py [hunk 8]
```diff
@@ -19,7 +19,7 @@ def run_test(self):
         def addr_checks(default_type):
             # Check default type, as well as explicit types
             types_and_addresses = [
-                (default_type, self.nodes[0].z_getnewaddress()),
+                (default_type, self.nodes[0].z_getnewaddress('sprout')),
```
Comment (LarryRuane) [inline qa/rpc-tests/wallet_addresses.py]: I don't think this is the right change; I think you should leave this line unchanged but modify the five callers to this function (`addr_checks()`) to pass `sapling` instead of `sprout`, because the argument is exactly the default type (which is changing)..

Comment (bitcartel) [reply qa/rpc-tests/wallet_addresses.py]: All the tests as they exist before this PR must explicitly ask for a sapling address, otherwise its the default sprout.  Therefore, any call to z_getnewaddress() intends to receive a sprout address.  Thats why I make the change here.  I also ran the full test suite locally and it passed.

File: src/test/rpc_wallet_tests.cpp (hunk 7 above)
Comment (bitcartel) [reply src/test/rpc_wallet_tests.cpp]: I notice that's the new idiom that is being used in some of the tests.  Please file a ticket and we can improve at a future date.

File: qa/rpc-tests/wallet_addresses.py (hunk 8 above)
Comment (daira) [reply qa/rpc-tests/wallet_addresses.py]: The comment is now wrong though, since this doesn't test the default. (There should be another test for the default.)

Comment (bitcartel) [reply qa/rpc-tests/wallet_addresses.py]: Ah, got it now.  Thanks.  Fixed as @LarryRuane described.  Also filed https://github.com/zcash/zcash/issues/3686 for @daira.

**Comment by bitcartel** [comment]
@Eirik0 @str4d Rebased and pushed to address comment.

File: src/test/rpc_wallet_tests.cpp (hunk 7 above)
Comment (LarryRuane) [reply src/test/rpc_wallet_tests.cpp]: Done, #3687. I'm not sure I understand what you mean about a new idiom; please update that issue if you think it's lacking.

**Comment by bitcartel** [comment]
@str4d Can you update review and once you're satisfied merge?  One thing is that even when Sapling has not been activated, the default z address returned will be a Sapling one.  So perhaps we should change the default based on epoch?  Or maybe it doesn't matter since Sapling has activated.

File: qa/rpc-tests/wallet_addresses.py [hunk 9]
```diff
@@ -39,39 +39,39 @@ def addr_checks(default_type):
         # Current height = 200 -> Sprout
         # Default address type is Sprout
```
Comment (str4d) [inline qa/rpc-tests/wallet_addresses.py]: Comment is now out-of-date.

File: qa/rpc-tests/wallet_addresses.py [hunk 10]
```diff
@@ -39,39 +39,39 @@ def addr_checks(default_type):
...
         print "Testing height 200 (Sprout)"
-        addr_checks('sprout')
+        addr_checks('sapling')
 
         self.nodes[0].generate(1)
         self.sync_all()
 
         # Current height = 201 -> Sprout
         # Default address type is Sprout
```
Comment (str4d) [inline qa/rpc-tests/wallet_addresses.py]: Comment is now out-of-date.

File: qa/rpc-tests/wallet_addresses.py [hunk 11]
```diff
@@ -39,39 +39,39 @@ def addr_checks(default_type):
...
         print "Testing height 201 (Sprout)"
-        addr_checks('sprout')
+        addr_checks('sapling')
 
         self.nodes[0].generate(1)
         self.sync_all()
 
         # Current height = 202 -> Overwinter
         # Default address type is Sprout
```
Comment (str4d) [inline qa/rpc-tests/wallet_addresses.py]: Comment is now out-of-date.

File: qa/rpc-tests/wallet_addresses.py [hunk 12]
```diff
@@ -39,39 +39,39 @@ def addr_checks(default_type):
...
         print "Testing height 202 (Overwinter)"
-        addr_checks('sprout')
+        addr_checks('sapling')
 
         self.nodes[0].generate(1)
         self.sync_all()
 
         # Current height = 203 -> Overwinter
         # Default address type is Sprout
```
Comment (str4d) [inline qa/rpc-tests/wallet_addresses.py]: Comment is now out-of-date.

File: qa/rpc-tests/wallet_addresses.py [hunk 13]
```diff
@@ -39,39 +39,39 @@ def addr_checks(default_type):
...
         print "Testing height 203 (Overwinter)"
-        addr_checks('sprout')
+        addr_checks('sapling')
 
         self.nodes[0].generate(1)
         self.sync_all()
 
         # Current height = 204 -> Sapling
         # Default address type is Sprout
```
Comment (str4d) [inline qa/rpc-tests/wallet_addresses.py]: Comment is now out-of-date.

File: qa/rpc-tests/wallet_addresses.py (hunk 9 above)
Comment (bitcartel) [reply qa/rpc-tests/wallet_addresses.py]: Added this to https://github.com/zcash/zcash/issues/3686

Comment (bitcartel) [reply qa/rpc-tests/wallet_addresses.py]: Scratch that, I will fix now.

**Comment by daira** [comment]
@bitcartel wrote:
> @str4d Can you update review and once you're satisfied merge? One thing is that even when Sapling has not been activated, the default z address returned will be a Sapling one. So perhaps we should change the default based on epoch? Or maybe it doesn't matter since Sapling has activated.

It does matter for forks that haven't yet activated Sapling. We should change the default based on the epoch at max(APPROX_RELEASE_HEIGHT, currentHeight).

**Comment by daira** [comment]
^ I no longer care about the above forks-that-haven't-activated-Sapling issue.
